utils/resb/src/binary.rs

```diff
@@ -0,0 +1,411 @@
+        Array = 8,
+
+        /// A 16-bit array resource.
+        ///
+        /// Consists of a 16-bit length value `n` followed by `n` 16-bit offsets
+        /// from the beginning of the 16-bit data block.
+        ///
+        /// As of [`FormatVersion::V3_0`], only `StringV2` representations can
+        /// be fully stored in the 16-bit data block. As such, only `StringV2`
+        /// resources can appear in an `Array16`.
+        Array16 = 9,
+
+        /// An integer array resource.
+        ///
+        /// Consists of a 32-bit length value `n` followed by `n` 32-bit integer
+        /// values.
+        ///
+        /// Note that these are not integer _resources_, but rather full 32-bit
+        /// integers.
+        IntVector = 14,
+    }
+);
+
+/// A `FormatVersion` represents a specific binary file format used for
+/// representing resource bundles.
+///
+/// A partial [specification] of each format version is present in the ICU4C
+/// source code.
+///
+/// [specification]: https://github.com/unicode-org/icu/blob/main/icu4c/source/common/uresdata.h
+#[derive(Clone, Copy, Debug, PartialEq, PartialOrd)]
+// We use `u32` representation to enforce correct sizing for structs containing
+// a `FormatVersion`.
+//
+// Note that order of declaration is important for partial ordering.
+#[repr(u32)]
+enum FormatVersion {
+    V1_0,
+    V1_1,
+    V1_2,
+    V1_3,
+    V2_0,
+    V3_0,
+}
+
+/// The `ResDescriptor` struct represents a typed pointer to a resource body
+/// within a binary resource bundle.
+///
+/// It is represented within the binary bundle as a 4-bit resource type in the
+/// most significant nibble of a 32-bit integer with a 28-bit unsigned offset
+/// in the remaining bits. The offset is interpreted as a count of 32-bit
+/// values from the start of the body.
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+struct ResDescriptor {
+    resource_type: ResourceReprType,
+    offset: u32,
+}
+
+impl ResDescriptor {
+    /// Makes a new resource descriptor with the given type and 28-bit offset.
+    pub const fn new(resource_type: ResourceReprType, offset: u32) -> Self {
+        Self {
+            resource_type,
+            offset,
+        }
+    }
+
+    /// Makes a new resource descriptor with the given type and no body.
+    pub const fn _new_empty(resource_type: ResourceReprType) -> Self {
+        Self::new(resource_type, 0)
+    }
+
+    /// Returns `true` if the described resource is empty.
+    pub fn is_empty(&self) -> bool {
+        self.offset == 0
+    }
+
+    /// Gets the 28-bit offset to the described resource.
+    pub fn offset(&self) -> u32 {
+        self.offset
+    }
+
+    /// Gets the resource type of the described resource.
+    pub fn resource_type(&self) -> ResourceReprType {
+        self.resource_type
+    }
+}
+
+/// The `Error` type provides basic error handling for deserialization of binary
+/// resource bundles.
+#[derive(Clone, Copy, Debug)]
+pub struct Error {
```

Comment:
Question: why did you make it `resb::binary::Error` instead of making a top-level type?

Reply:
The style guide dictates that you should also name it something more specific like `BinaryError`.

Reply:
Each feature (binary serializer, binary deserializer, and text parser) has its own error type specific to the errors that it can produce. I've renamed them appropriately.